dace/transformation/interstate/gpu_transform_sdfg.py

```diff
@@ -192,27 +192,196 @@ def apply(self, _, sdfg: sd.SDFG):
+                        scalset = scalset.union(sset)
+                        scalout = scalout and ssout
+                        continue
+                    if desc.storage not in gpu_storage and last_edge.data.num_elements() == 1:
+                        sset, ssout = _recursive_in_check(last_edge.src, state, gpu_scalars)
+                        scalset = scalset.union(sset)
+                        scalout = scalout and ssout
+                        continue
+                    scalout = False
+            return scalset, scalout
+
+        def _is_devicelevel_gpu(sdfg, state, node):
+            """ Recursive version of `is_devicelevel_gpu` method. """
+            # Returns True only for nodes that are actually on a GPU scope.
+            # This excludes NestedSDFGs that have a GPU-Device schedule but are not in a GPU kernel.
+            is_parent_nested = (sdfg.parent is not None)
+            if is_parent_nested:
+                return scope.is_devicelevel_gpu(sdfg.parent.parent,
+                                                sdfg.parent,
+                                                sdfg.parent_nsdfg_node,
+                                                with_gpu_default=True)
+            else:
+                return scope.is_devicelevel_gpu(state.parent, state, node, with_gpu_default=True)
+
+        gpu_scalars = {}
+        # Iterates over Tasklets that not inside a GPU kernel. Such Tasklets must be moved inside a GPU kernel only
+        # if they write to GPU memory. The check takes into account the fact that GPU kernels can read host-based
+        # Scalars, but cannot write to them.
+        for node, state in sdfg.all_nodes_recursive():
+            if isinstance(node, nodes.Tasklet):
+                if state.entry_node(node) is None and not _is_devicelevel_gpu(state.parent, state, node):
+                    scalars, scalar_output = _recursive_out_check(node, state, gpu_scalars)
+                    sset, ssout = _recursive_in_check(node, state, gpu_scalars)
+                    scalars = scalars.union(sset)
+                    scalar_output = scalar_output and ssout
+                    csdfg = state.parent
+                    # If the tasklet is not adjacent only to scalars or it is in a GPU scope.
+                    # The latter includes NestedSDFGs that have a GPU-Device schedule but are not in a GPU kernel.
+                    if (not scalar_output or (csdfg.parent is not None
+                                              and csdfg.parent_nsdfg_node.schedule == dtypes.ScheduleType.GPU_Default)):
+                        global_code_nodes[state].append(node)
+                        gpu_scalars.update({k: None for k in scalars})
+
+        # NOTE: We execute the above algorithm a second time to catch any potential new GPU-scalars.
+        # TODO: Should we run this recursively until there are no new GPU-scalars?
```

Comment:
can we simply run this once, but reversed? Since new gpu scalars only affect tasklets if they are output, we only need it to run once, or am I wrong?

Reply:
Data can have access nodes in multiple states. Therefore, if some data is moved to GPU memory in one state, then we need to rerun the check for all other states that is used in and have been already checked. There is probably a better way to handle this and this is why I have left {NOTE, TODO}.

Reply:
Agreed.